src/components/TodoItem.tsx

```diff
@@ -0,0 +1,177 @@
+  const loaders = todo.id === 0
+  || (todo.id === todoId && checkedLoading)
+  || (todo.completed && checkedLoad)
+  || (!todo.completed && toggle === StatusFilter.COMPLETED)
+  || (todo.completed && toggle === StatusFilter.ACTIVE);
+
+  const todoItemEditRef = useRef<HTMLInputElement | null>(null);
+
+  useEffect(() => {
+    if (switchEditTodo) {
+      todoItemEditRef.current?.focus();
+    }
+  }, [switchEditTodo]);
+
+  const handleDelete = (tId: number) => {
+    setCheckedLoading(true);
+    onDeleteTodo(tId)
+      .finally(() => {
+        fieldTitle.current?.focus();
+        setCheckedLoading(false);
+      });
+  };
+
+  const handleChecked = (
+    tod: Todo,
+    e: React.ChangeEvent<HTMLInputElement>,
+  ) => {
+    setCheckedLoading(true);
+    onCheckedTodo({ ...tod, completed: e.target.checked })
+      .finally(() => setCheckedLoading(false));
+  };
+
+  const handleUpdateTitle = (tod: Todo) => {
+    if (!newTitle.trim()) {
+      setCheckedLoading(true);
+      onDeleteTodo(tod.id)
+        .then(() => setSwitchEditTodo(false))
+        .catch(() => todoItemEditRef.current?.focus())
+        .finally(() => setCheckedLoading(false));
+    } else {
+      setCheckedLoading(true);
+      updateTitle({ ...tod, title: newTitle.trim() })
+        .then(() => setSwitchEditTodo(false))
+        .catch(() => todoItemEditRef.current?.focus())
+        .finally(() => setCheckedLoading(false));
+    }
+  };
+
+  const handlesForUpdateTitle = () => {
+    setNewTitle(newTitle);
+    handleUpdateTitle(todo);
+  };
+
+  const handlesCurrentTitle = () => {
+    setNewTitle(todo.title);
+    setSwitchEditTodo(false);
+  };
+
+  const handleKeyUp = (e: React.KeyboardEvent) => {
+    if (e.key === 'Enter') {
+      if (newTitle === todo.title) {
+        handlesCurrentTitle();
+      } else {
+        handlesForUpdateTitle();
+      }
+    } else if (e.key === 'Escape') {
+      handlesCurrentTitle();
+    }
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={todo.completed ? 'todo completed' : 'todo'}
+      key={todo.id}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          onChange={(e) => {
+            handleChecked(todo, e);
+            setTodoId(todo.id);
+          }}
+          checked={todo.completed}
+        />
+      </label>
+
+      {
+        switchEditTodo ? (
+          <input
+            data-cy="TodoTitleField"
+            type="text"
+            className="todo__title-field"
+            value={newTitle}
+            ref={todoItemEditRef}
+            onChange={(e) => {
+              setNewTitle(e.target.value);
+            }}
+            onBlur={() => {
+              if (newTitle === todo.title) {
+                handlesCurrentTitle();
+              } else {
+                handlesForUpdateTitle();
+              }
+            }}
```

Comment:
Consider will be better if you move this logic to the helper function and use it here